lib/Target/ExportSystemC/Patterns/BuiltinEmissionPatterns.cpp

```diff
@@ -76,6 +76,36 @@ struct IndexTypeEmitter : TypeEmissionPattern<IndexType> {
   void emitType(IndexType type, EmissionPrinter &p) override { p << "size_t"; }
 };
 
+/// Emit a builtin function type as a std::function.
+struct FunctionTypeEmitter : public TypeEmissionPattern<FunctionType> {
+  bool match(Type type) override {
+    return type.isa<FunctionType>() &&
+           type.cast<FunctionType>().getNumResults() <= 1;
+  }
+
+  void emitType(FunctionType type, EmissionPrinter &p) override {
+    p << "std::function<";
+
+    if (type.getNumResults() == 0)
+      p << "void";
+    else
+      p.emitType(type.getResult(0));
+
+    p << "(";
+
+    bool first = true;
+    for (Type ty : type.getInputs()) {
+      if (!first)
+        p << ", ";
+
+      p.emitType(ty);
+      first = false;
+    }
```

Comment:
Probably you can simplify this with `llvm::interleaveComma` 
```suggestion
   llvm::interleaveComma(type.getInputs(), p,
                          [&](auto ty) { p.emitType(ty); });
```